pallets/gear/src/migrations.rs

```diff
@@ -0,0 +1,311 @@
+                if random() { Some(random()) } else { None },
+            ))
+        };
+        StoredDispatch::new(
+            kind,
+            StoredMessage::new(
+                MessageId::from(random::<u64>()),
+                ProgramId::from(random::<u64>()),
+                destination,
+                random_payload(),
+                random(),
+                details,
+            ),
+            context,
+        )
+    }
+
+    #[test]
+    fn migration_works() {
+        init_logger();
+
+        new_test_ext().execute_with(|| {
+            StorageVersion::new(3).put::<GearProgram>();
+
+            let multiplier = <Test as pallet_gear_bank::Config>::GasMultiplier::get();
+
+            let destinations: Vec<ProgramId> =
+                up_to(32, || ProgramId::from(random::<u64>())).collect();
+            let dispatches: Vec<_> = destinations
+                .iter()
+                .cloned()
+                .map(|destination| {
+                    (
+                        destination,
+                        up_to(32, || random_dispatch(destination)).collect::<Vec<_>>(),
+                    )
+                })
+                .collect();
+
+            for (destination, dispatches) in dispatches.clone() {
+                let mut messages = vec![];
+                for dispatch in dispatches {
+                    messages.push(dispatch.id());
+
+                    GasHandlerOf::<Test>::create(USER_1, multiplier, dispatch.id(), 0).unwrap();
+                    WaitlistOf::<Test>::insert(
+                        dispatch,
+                        BlockNumberFor::<Test>::from(random::<u64>()),
+                    )
+                    .unwrap();
+                }
+
+                waiting_init_list::WaitingInitStorage::<Test>::insert(destination, messages);
+            }
+
+            let state = MigrateWaitingInitList::<Test>::pre_upgrade().unwrap();
+            let weight = MigrateWaitingInitList::<Test>::on_runtime_upgrade();
+            assert!(!weight.is_zero());
+            MigrateWaitingInitList::<Test>::post_upgrade(state).unwrap();
+
+            assert_eq!(StorageVersion::get::<GearProgram>(), 4);
+
+            assert_eq!(
+                waiting_init_list::WaitingInitStorage::<Test>::iter().count(),
+                0
+            );
+
+            let messages_in_queue: usize = dispatches.iter().map(|(_, d)| d.len()).sum();
+            assert_eq!(QueueOf::<Test>::iter().count(), messages_in_queue);
+
+            let mut compared = 0;
+            for (destination, dispatches) in dispatches.clone() {
+                for dispatch in dispatches {
+                    let LinkedNode {
+                        value: queued_dispatch,
+                        ..
+                    } = Dispatches::<Test>::get(dispatch.id()).unwrap();
+                    assert_eq!(queued_dispatch, dispatch);
+                    assert_eq!(queued_dispatch.destination(), destination);
```

Comment:
for test purposes you may run queue here to assert errors in logs, if we've chosen pallet gear for migration place